### ISSUE TITLE
setting KUBE_VPC_CIDR_BASE breaks kube-up.sh for AWS

### ISSUE THREAD

**Issue by captify-akraievoy** [kubernetes/kubernetes#33559]
**Is this a BUG REPORT or FEATURE REQUEST?**: BUG REPORT

**Kubernetes version**:
`Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:16:57Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/386"}
`

**Environment**:
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release):
```
NAME="Ubuntu"
VERSION="14.04.3 LTS, Trusty Tahr"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 14.04.3 LTS"
VERSION_ID="14.04"
```
- **Kernel** (e.g. `uname -a`):
'Linux ak 3.16.0-41-generic #57~14.04.1-Ubuntu SMP Thu Jun 18 18:01:13 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux'
- **Install tools**: Just downloaded and extracted release tar.gz (1.3.6 was installed the same way)
- **Others**:
This is kinda blocker for us, because the default VPC CIDR is already taken by dev k8s cluster and we have to have VPC peering to both dev and prod k8s VPCs. So the CIDRs must be different.

**What happened**:
As soon as I attempt to supply property `KUBE_VPC_CIDR_BASE` to cube-up.sh - it does not complete. Master node accepts SSH connections but does not accept HTTPS connections at all.

The script I use to run kube-up.sh:
```
#!/bin/sh

export KUBERNETES_PROVIDER=aws
export KUBE_AWS_ZONE=eu-west-1c
export KUBE_VPC_CIDR_BASE=172.40
export NUM_NODES=3
export MASTER_SIZE=m3.large
export NODE_SIZE=m4.2xlarge
export AWS_S3_REGION=eu-west-1
export AWS_S3_BUCKET=mycompany-k8s-prod-artifacts
export KUBE_AWS_INSTANCE_PREFIX=k8s-prod-140
export AWS_SSH_KEY=$HOME/.ssh/kubernetes-prod.pem
export MASTER_DISK_SIZE=64
export MASTER_ROOT_DISK_SIZE=32
export NODE_ROOT_DISK_SIZE=64

kubernetes/cluster/kube-up.sh

```

And the respective log:

```
kubernetes/cluster/kube-up.sh
... Starting cluster in eu-west-1c using provider aws
... calling verify-prereqs
... calling kube-up
Starting cluster using os distro: jessie
Uploading to Amazon S3
+++ Staging server tars to S3 Storage: mycompany-k8s-prod-artifacts/devel
upload: ../../tmp/kubernetes.O0yWx0/s3/bootstrap-script to s3://mycompany-k8s-prod-artifacts/devel/bootstrap-script
Uploaded server tars:
  SERVER_BINARY_TAR_URL: https://s3-eu-west-1.amazonaws.com/mycompany-k8s-prod-artifacts/devel/kubernetes-server-linux-amd64.tar.gz
  SALT_TAR_URL: https://s3-eu-west-1.amazonaws.com/mycompany-k8s-prod-artifacts/devel/kubernetes-salt.tar.gz
  BOOTSTRAP_SCRIPT_URL: https://s3-eu-west-1.amazonaws.com/mycompany-k8s-prod-artifacts/devel/bootstrap-script
INSTANCEPROFILE	arn:aws:iam::412792291881:instance-profile/kubernetes-master	2016-08-01T12:56:02Z	AIPAJ4K2B26XVWY4M2NKY	kubernetes-master	/
ROLES	arn:aws:iam::412792291881:role/kubernetes-master	2016-08-01T12:55:59Z	/	AROAJXQX5WZX3WLGHPQNC	kubernetes-master
ASSUMEROLEPOLICYDOCUMENT	2012-10-17
STATEMENT	sts:AssumeRole	Allow
PRINCIPAL	ec2.amazonaws.com
INSTANCEPROFILE	arn:aws:iam::412792291881:instance-profile/kubernetes-minion	2016-08-01T12:56:07Z	AIPAJYI3XUE3I25A7ZGPA	kubernetes-minion	/
ROLES	arn:aws:iam::412792291881:role/kubernetes-minion	2016-08-01T12:56:05Z	/	AROAI3EEVZI2TRFSM36NK	kubernetes-minion
ASSUMEROLEPOLICYDOCUMENT	2012-10-17
STATEMENT	sts:AssumeRole	Allow
PRINCIPAL	ec2.amazonaws.com
Using SSH key with (AWS) fingerprint: b8:a4:7a:32:44:4d:43:80:e8:a9:f6:2e:8b:ef:14:71
Creating vpc.
Adding tag to vpc-16e07d72: Name=kubernetes-vpc
Adding tag to vpc-16e07d72: KubernetesCluster=k8s-prod-140
Using VPC vpc-16e07d72
Adding tag to dopt-31e8fa54: Name=kubernetes-dhcp-option-set
Adding tag to dopt-31e8fa54: KubernetesCluster=k8s-prod-140
Using DHCP option set dopt-31e8fa54
Creating subnet.
Adding tag to subnet-49398411: KubernetesCluster=k8s-prod-140
Using subnet subnet-49398411
Creating Internet Gateway.
Using Internet Gateway igw-b7f7ecd2
Associating route table.
Creating route table
Adding tag to rtb-01f3aa65: KubernetesCluster=k8s-prod-140
Associating route table rtb-01f3aa65 to subnet subnet-49398411
Adding route to route table rtb-01f3aa65
Using Route Table rtb-01f3aa65
Creating master security group.
Creating security group kubernetes-master-k8s-prod-140.
Adding tag to sg-007f9c66: KubernetesCluster=k8s-prod-140
Creating minion security group.
Creating security group kubernetes-minion-k8s-prod-140.
Adding tag to sg-087f9c6e: KubernetesCluster=k8s-prod-140
Using master security group: kubernetes-master-k8s-prod-140 sg-007f9c66
Using minion security group: kubernetes-minion-k8s-prod-140 sg-087f9c6e
Creating master disk: size 64GB, type gp2
Adding tag to vol-745360c5: Name=k8s-prod-140-master-pd
Adding tag to vol-745360c5: KubernetesCluster=k8s-prod-140
Allocated Elastic IP for master: 52.30.198.167
Adding tag to vol-745360c5: kubernetes.io/master-ip=52.30.198.167
Generating certs for alternate-names: IP:52.30.198.167,IP:172.40.0.9,IP:10.0.0.1,DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local,DNS:k8s-prod-140-master
Starting Master
Adding tag to i-c49e7c4b: Name=k8s-prod-140-master
Adding tag to i-c49e7c4b: Role=k8s-prod-140-master
Adding tag to i-c49e7c4b: KubernetesCluster=k8s-prod-140
Waiting for master to be ready
Attempt 1 to check for master nodeWaiting for instance i-c49e7c4b to be running (currently pending)
Sleeping for 3 seconds...
Waiting for instance i-c49e7c4b to be running (currently pending)
Sleeping for 3 seconds...
Waiting for instance i-c49e7c4b to be running (currently pending)
Sleeping for 3 seconds...
 [master running]
Attaching IP 52.30.198.167 to instance i-c49e7c4b
Attaching persistent data volume (vol-745360c5) to master
2016-09-27T12:57:59.886Z	/dev/sdb	i-c49e7c4b	attaching	vol-745360c5
cluster "aws_k8s-prod-140" set.
user "aws_k8s-prod-140" set.
context "aws_k8s-prod-140" set.
switched to context "aws_k8s-prod-140".
user "aws_k8s-prod-140-basic-auth" set.
Wrote config for aws_k8s-prod-140 to /home/akraievoy/.kube/config
Creating minion configuration
Creating autoscaling group
 0 minions started; waiting
 0 minions started; waiting
 0 minions started; waiting
 0 minions started; waiting
 3 minions started; ready
Waiting for cluster initialization.

  This will continually check to see if the API for kubernetes is reachable.
  This might loop forever if there was some uncaught error during start
  up.

......................................................................................................
```

I am able to open shell to the master node, like so `ssh -i .ssh/kubernetes-prod.pem admin@52.30.198.167`, then I sift through /var/log/syslog - and find something like the below in it, which looks like some sort of provisioning problem:

```
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: [ERROR   ] Unable to manage file: Jinja variable list object has no element 0; line 12
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: ---
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: [...]
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: {% set api_servers = "--api-servers=https://" + grains.api_servers -%}
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: {% elif grains.apiservers is defined -%} # TODO(remove after 0.16.0): Deprecated form
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: {% set api_servers = "--api-servers=https://" + grains.apiservers -%}
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: {% elif grains['roles'][0] == 'kubernetes-master' -%}
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: {% set master_ipv4 = salt['grains.get']('fqdn_ip4')[0] -%}
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: {% set api_servers = "--api-servers=https://" + master_ipv4 -%}    <======================
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: {% else -%}
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: {% set ips = salt['mine.get']('roles:kubernetes-master', 'network.ip_addrs', 'grain').values() -%}
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: {% set api_servers = "--api-servers=https://" + ips[0][0] -%}
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: {% endif -%}
Sep 27 13:03:33 ip-172-40-0-9 rc.local[374]: [...]
```

Then there is something like provisioning summary several screenfuls away from that crashsite which looks like 
```
Sep 27 13:03:36 ip-172-40-0-9 rc.local[374]: -------------
Sep 27 13:03:36 ip-172-40-0-9 rc.local[374]: Succeeded: 89
Sep 27 13:03:36 ip-172-40-0-9 rc.local[374]: Failed:     6
Sep 27 13:03:36 ip-172-40-0-9 rc.local[374]: -------------
Sep 27 13:03:36 ip-172-40-0-9 rc.local[374]: Total:     95
```

**What you expected to happen**:
Well I had a full log of kube-up.sh completing successfully as soon as I comment the KUBE_VPC_CIDR_BASE property - but now I think it's uninteresting.

Also I have verified that the above particular provisioning problem does *not* occur in master's /var/log/syslog (for the case when property KUBE_VPC_CIDR_BASE is *not* set) - and indeed it does not:
```
Sep 27 13:33:03 ip-172-20-0-9 rc.local[375]: Summary
Sep 27 13:33:03 ip-172-20-0-9 rc.local[375]: -------------
Sep 27 13:33:03 ip-172-20-0-9 rc.local[375]: Succeeded: 95
Sep 27 13:33:03 ip-172-20-0-9 rc.local[375]: Failed:     0
Sep 27 13:33:03 ip-172-20-0-9 rc.local[375]: -------------
Sep 27 13:33:03 ip-172-20-0-9 rc.local[375]: Total:     95
```

**How to reproduce it**:
Set KUBE_VPC_CIDR_BASE=172.40 - all other options I'm setting do not break kube-up.sh
After my tinkering around I came to opinion that this option is very likely to nuke your rollout even if you keep all other values to respective sensible defaults.

**Anything else do we need to know**:
I have played around with this: 
 * attempting to wipe my .kube/config
 * changing cluster id (thought some previous kube-up is somehow interfering) 
 * installing from both my physical office machine and another amazon cloud instance (thought some routing misconfiguration was affecting my install)

**Comment by captify-akraievoy** [comment]
@justinsb distress call here: please point me for further investigation directions

**Comment by captify-akraievoy** [comment]
I was able to un-puzzle the whole thing myself, so the issue should be closed. Maybe some minor bit of documentation could be added for such illiterate folks like me - so that they won't do the same mistake.

I had a suspicion that the problem was caused by private DNS names not resolving properly, which https://github.com/saltstack/salt/issues/34129 also was suggesting.

Trying to see whether this was actually the case:
The new cluster refuses to resolve local host name:
```
admin@ip-172-40-1-9:~$ nslookup ip-172-40-1-9
Server:		172.40.0.2
Address:	172.40.0.2#53

** server can't find ip-172-40-1-9: NXDOMAIN
```
While previous _default_ VPC master _does_ resolve local host name:
```
root@ip-172-20-0-9:/home/admin# nslookup ip-172-20-0-9
Server:		172.20.0.2
Address:	172.20.0.2#53

Non-authoritative answer:
Name:	ip-172-20-0-9.eu-west-1.compute.internal
Address: 172.20.0.9

```

I then decided to double-check what Amazon recommends in documentation, and found the below [here](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/vpc-dns.html):

> The Amazon DNS server cannot resolve private DNS hostnames if your VPC's IP address range falls outside of the private IP addresses ranges specified by RFC 1918.

So, maybe adding a brief remark into [cluster/aws/options.md](https://github.com/kubernetes/kubernetes/blame/5ead89b/cluster/aws/options.md#L151) that KUBE_VPC_CIDR_BASE property should be within 172.16..172.31 range may prove helpful for someone else.
